pkg/model/migrator/receipt.go

```diff
@@ -0,0 +1,247 @@
+package migrator
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/gohornet/hornet/pkg/model/utxo"
+	"github.com/iotaledger/iota.go/encoding/t5b1"
+	iotago "github.com/iotaledger/iota.go/v2"
+)
+
+const (
+	receiptFilePattern = "%d.%d.json"
+)
+
+var (
+	// Returned when the state of the ReceiptService is invalid.
+	ErrInvalidReceiptServiceState = errors.New("invalid receipt service state")
+)
+
+// ReceiptPersistFunc is a function which persists a receipt.
+type ReceiptPersistFunc func(r *iotago.Receipt) error
+
+// ReceiptValidateFunc is a function which validates a receipt.
+type ReceiptValidateFunc func(r *iotago.Receipt) error
+
+// ReceiptService is in charge of persisting and validating a batch of receipts.
+type ReceiptService struct {
+	// Whether the service is configured to back up receipts.
+	BackupEnabled bool
+	// Whether the service is configured to validate receipts.
+	ValidationEnabled bool
+	// Whether the service should ignore soft errors.
+	IgnoreSoftErrors bool
+	backupFolder     string
+	validator        *Validator
+	utxoManager      *utxo.Manager
+}
+
+// NewReceiptService creates a new ReceiptService.
+func NewReceiptService(v *Validator, utxoManager *utxo.Manager, validationEnabled bool, backupEnabled bool, ignoreSoftErrors bool, backupFolder string) *ReceiptService {
+	return &ReceiptService{
+		ValidationEnabled: validationEnabled,
+		IgnoreSoftErrors:  ignoreSoftErrors,
+		BackupEnabled:     backupEnabled,
+		utxoManager:       utxoManager,
+		validator:         v,
+		backupFolder:      backupFolder,
+	}
+}
+
+// Init initializes the ReceiptService and returns the amount of receipts currently stored.
+func (rs *ReceiptService) Init() error {
+	if !rs.BackupEnabled {
+		return nil
+	}
+	if err := os.MkdirAll(rs.backupFolder, os.ModePerm); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Backup backups the given receipt to disk.
+func (rs *ReceiptService) Backup(r *utxo.ReceiptTuple) error {
+	if !rs.BackupEnabled {
+		panic("receipt service is not configured to backup receipts")
+	}
+
+	receiptFileName := path.Join(rs.backupFolder, fmt.Sprintf(receiptFilePattern, r.Receipt.MigratedAt, r.MilestoneIndex))
+	receiptJSON, err := r.Receipt.MarshalJSON()
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(receiptFileName, receiptJSON, os.ModePerm); err != nil {
+		return fmt.Errorf("unable to persist receipt onto disk: %w", &SoftError{Err: err})
```

Comment:
why is not persisting the backup a SoftError?

Reply:
I believe you're right that the argument can be made, that if someone enabled backups of receipts, then they probably want the node to go hi-wire in case they can't be persisted. Going to change it to a crit err